### PR TITLE
Allow target value updates to be skipped

### LINKED ISSUE #25797: WindowCovering: hard to implement StopMotion Command for hardware.
### Reproduction steps

The current implemention of [emberAfWindowCoveringClusterStopMotionCallback](https://github.com/project-chip/connectedhomeip/blob/bd94d060c3656c76bfd410f20698639bda68cc1e/src/app/clusters/window-covering-server/window-covering-server.cpp#L721) causes the attribute value of current value to be inconsistent with the actual value.

**What happenned**
As the spec saids, the StopMotion command shall:

- TargetPositionLiftPercent100ths attribute will be set to CurrentPositionLiftPercent100ths attribute value.
- TargetPositionTiltPercent100ths attribute will be set to CurrentPositionTiltPercent100ths attribute value.

[emberAfWindowCoveringClusterStopMotionCallback](https://github.com/project-chip/connectedhomeip/blob/bd94d060c3656c76bfd410f20698639bda68cc1e/src/app/clusters/window-covering-server/window-covering-server.cpp#L721) get the value of `CurrentPositionLiftPercent100ths` and set the value to `TargetPositionLiftPercent100ths` immediately.

But some hardwares update `CurrentPositionLiftPercent100ths` periodically. The value of `CurrentPositionLiftPercent100ths` may differ than the real position.

For example, the hardware design may put `connectedhomeip` on a high performance processor with network capabilities, and put motor control on another low performance controller. The two communicate using UART.

Assuming that the target value is now set to 100%, current value is set to 0%. When current value increases to 50%, the application invokes StopMotion. Then the timeline and status of each value are as follows:

```
|          time |   0   |  10  |  20  |  30  |  40  |  50  |  51  |
|         motor |   0%  |  10% |  20% |  30% |  40% |  50% |  51% |
|   (UART Comm.)| *SYNC |      |*SYNC |      |*SYNC |*STOP |*SYNC?|
| Current Value |   0%  |   0% |  20% |  20% |  40% |  40% |  51%?|
|  Target Value | 100%  | 100% | 100% | 100% | 100% |  40% |  40%?|
```
^ *SYNC: motor controller reports its current value to `connectedhomeip`.
^ *STOP: `connectedhomeip` send STOP command to motor.

At the time 50, StopMotion is received. `emberAfWindowCoveringClusterStopMotionCallback` get the current value base on the last sync value, and it set the target value to the out-dated value. And this results in the target value and current value not as expected (which is 50%).

### Bug prevalence

--

### GitHub hash of the SDK that was being used

master/v1.0-branch

### Platform

core

### Platform Version(s)

_No response_

### Anything else?

**prefer solution**

There are some solutions I have considered:
1. the motor stop work immediately, and **do not update** current value to its real current value. 
   This solution will result the value on attribute and the real value mismatched.
2. the motor stop work immediately, and **update** current value to its real value after `emberAfWindowCoveringClusterStopMotionCallback`
   This solution will result current value and target value mismatched.
3. the motor stop and **move back** to 40%.
   This solution will result "StopMotion" does not acturally stop the motor. And mostly it is not what users want.
4. **block working thread** in HandleStopMotion() delegate, and wait motor sync its current value. After that, finish `emberAfWindowCoveringClusterStopMotionCallback`.
   This solution will result in nearly all operations stop working for an undetermined amount of time. And the code in HandleStopMotion() can not reply error status to the invoker, if motor controller does not report its states in time.
5. **update target value** when motor update its real current value, after `emberAfWindowCoveringClusterStopMotionCallback`
   This solution will result StopMotion **does not work as the spec expected** slightly. When StopMotion response to the command invoker, the target value does not changed immediately. Instead, it changed slightly later.
6. make `emberAfWindowCoveringClusterStopMotionCallback` could handle command asynchronously. And then, we could wait for the motor controller report its state without blocking working thread. After that, send command response.
   This solution requires modification on the processing flow of InvokeRequest. And it requires more discussion.

I prefer solution 4 currently. But I expect the solution 6 to be discussed, and be implmemented in `connectedhomeip`. I think it will be a better solution than solution 4 if it is implmeneted.

Please let me know if there are other solutions. Thank you!

### PR DESCRIPTION
**Pull request by Kxuan**
Fixes #25797

This patch: 

* Added CHIP_ERROR_INPROGRESS error code
* If delegate->HandleStopMotion returns CHIP_ERROR_INPROGRESS, the target value will not be set.

The attributes "TargetPositionLiftPercent100ths" and "TargetPositionTiltPercent100ths" will be updated to the attribute value of their coresponding current values.

But the attribute value and the real state may be differ. It may have a delay in updating the real state to the attribute value. It results in the target value mismatched with the real value we want.

emberAfWindowCoveringClusterStopMotionCallback set the target value to an out-dated value first. And then, when the hardware reports its state later, the corrected value will be updated again.
The subscribers will receive two attribute "Report", the first one has an incorrect value, and the second one has the correct value.

This patch fix this issue by allowing applications skip the first update on the target value.

When the application code wants to update the target value by itself, it can return CHIP_ERROR_INPROGRESS in HandleStopMotion. Then it updates the target value later.

And then the first incorrect value will no longer be reported to subscribers.

### CODE REVIEW

File: src/lib/core/CHIPError.h [hunk 1]
```diff
@@ -2435,6 +2435,14 @@ using CHIP_ERROR = ::chip::ChipError;
  */
 #define CHIP_ERROR_HANDLER_NOT_SET                            CHIP_CORE_ERROR(0xe1)
 
+/**
+ * @def CHIP_ERROR_INPROGRESS
+ *
+ * @brief
+ *   The operation is still in progress
+ */
+#define CHIP_ERROR_INPROGRESS                                 CHIP_CORE_ERROR(0xe2)
```
Comment (bzbarsky-apple) [inline src/lib/core/CHIPError.h:2444]: There's an existing gap at 0xa4.  Could you please use that value instead of adding this at the end?

Comment (bzbarsky-apple) [reply src/lib/core/CHIPError.h:2444]: Also, docs/ERROR_CODES.md needs to be updated, right?

Comment (Kxuan) [reply src/lib/core/CHIPError.h]: Thank you @bzbarsky-apple 
I have updated its value to 0xa4 and updated the table in ERROR_CODES.md.

- Please take a look.

Comment (jmeg-sfy) [reply src/lib/core/CHIPError.h]: @bzbarsky-apple @Kxuan , does CHIP_ERROR_INPROGRESS is in the specs ?
i am not against it but maybe the specs need to be updated following this behavior

File: src/lib/core/CHIPError.h [hunk 2]
```diff
@@ -1372,7 +1372,14 @@ using CHIP_ERROR = ::chip::ChipError;
...
 // AVAILABLE: 0xa3
-// AVAILABLE: 0xa4
+
+/**
+ * @def CHIP_ERROR_INPROGRESS
+ *
+ * @brief
+ *   The operation is still in progress
+ */
+#define CHIP_ERROR_INPROGRESS                                 CHIP_CORE_ERROR(0xa4)
```
Comment (jmeg-sfy) [inline src/lib/core/CHIPError.h:1382]:
```suggestion
#define CHIP_ERROR_IN_PROGRESS                                 CHIP_CORE_ERROR(0xa4)
```

File: src/lib/core/CHIPError.cpp [hunk 3]
```diff
@@ -347,6 +347,9 @@ bool FormatCHIPError(char * buf, uint16_t bufSize, CHIP_ERROR err)
     case CHIP_ERROR_IM_FABRIC_DELETED.AsInteger():
         desc = "The fabric is deleted, and the corresponding IM resources are released";
         break;
+    case CHIP_ERROR_INPROGRESS.AsInteger():
```
Comment (jmeg-sfy) [inline src/lib/core/CHIPError.cpp:350]:
```suggestion
    case CHIP_ERROR_IN_PROGRESS.AsInteger():
```

File: src/lib/core/CHIPError.h (hunk 1 above)
Comment (bzbarsky-apple) [reply src/lib/core/CHIPError.h]: @jmeg-sfy CHIP_ERROR values do not generally have anything to do with specs...

File: src/app/clusters/window-covering-server/window-covering-server.cpp [hunk 4]
```diff
@@ -735,26 +735,39 @@ bool emberAfWindowCoveringClusterStopMotionCallback(app::CommandHandler * comman
...
+            LogErrorOnFailure(err);
+        }
     }
     else
     {
         emberAfWindowCoveringClusterPrint("WindowCovering has no delegate set for endpoint:%u", endpoint);
     }
 
-    if (HasFeaturePaLift(endpoint))
+    if (changeTarget)
```
[the remark this answers is not in the document]
Comment (Kxuan) [reply src/app/clusters/window-covering-server/window-covering-server.cpp:758]: The code that returns the `CHIP_ERROR_IN_PROGRESS` error code is responsible for updating these attributes' value.
Normally, it is the manufacturor's code.

When the actural value have to be read from an external device, the manufacturor's code can send a request to the external device and return `CHIP_ERROR_IN_PROGRESS` to CHIP. Then, when the external device stops the motor and reports its actural state, the manufacturor's code updates the actural value to CHIP.

It prevents CHIP report out-dated value to the attribute subscribers.

**Comment by Kxuan** [comment]
Thanks @jmeg-sfy . I have renamed the name of the error code to "CHIP_ERROR_IN_PROGRESS". PTAL

**Review by andy31415 — APPROVED** [review]

**Review by jmeg-sfy — APPROVED** [review]

**Pull request merged**